apps/interpret_nn/halide/common_halide.cpp

```diff
@@ -78,12 +79,10 @@ Expr round_shift_right(const Expr &x, const Expr &exponent) {
     Type t = x.type();
     Type t_unsigned = t.with_code(halide_type_uint);
     Expr uexponent = cast(t_unsigned, exponent);
-    // Exponent must satisfy 0 <= exponent <= 31
-    // TODO: Maybe this should be an offset added to x prior to shifting.
-    Expr mask = (cast(x.type(), 1) << uexponent) - 1;
-    Expr remainder = x & mask;
-    Expr threshold = (mask >> 1) + (x < 0);
-    return (x >> uexponent) + (remainder > threshold);
+    Expr one = cast(x.type(), 1);
+    // This is intended to pattern-match against rounding-shift-right instruction generation.
+    Expr rounding = (one << uexponent) >> 1;
+    return (saturating_cast(t, x + rounding) >> uexponent);
```

Comment:
I think this isn't quite a saturating add, you would need to widen x first. Something like:

    saturating_cast(t, cast(x.type().widen(), x) + rounding)

It's a bit messy for generic type code :( It's better if we have a `widen` helper function, and a `saturating_narrow` helper doesn't hurt either:

    return saturating_narrow(widen(x) + rounding) >> uexponent;

Reply:
...why don't we have explicit helpers for these in (eg) IROperators.h, in addition to pattern matching?

Reply:
PTAL

Reply:
> ...why don't we have explicit helpers for these in (eg) IROperators.h, in addition to pattern matching?

I see that these are being added in https://github.com/halide/Halide/pull/5531 -- woot

Reply:
They're being added, but not that in that branch, they're still in the "Internal" namespace, and shouldn't be used directly. Promoting them is something we can talk about too, but in a separate change of course.